ankimorphs/highlighting/text_highlighter.py

```diff
@@ -0,0 +1,388 @@
+             is incorrect, we cannot make an inference on how well the text is known.
+
+        8. Ruby starts, then status starts, ruby ends, and status ends:
+            Example: '謎解[なぞと]き'
+            Action: wrap everything with the status ankimorphs_globals.STATUS_UNDEFINED
+            Explanation: this is a combination of scenarios 6 and 7.
+             The correct word here is '謎解き' (scenario 6), but the morphemizer splits it into ['謎','解き'] (scenario 7).
+        """
+
+        ruby: Ruby | None = None
+        status: Status | None = None
+
+        while_counter = -1
+
+        while self._highlighted_expression is not None:
+
+            debug_utils.dev_print(f"self._highlighted: {self._highlighted_expression}")
+
+            while_counter += 1
+            debug_utils.dev_print(f"while counter: {while_counter}")
+
+            if ruby is None and self.rubies:
+                ruby = self.rubies.pop()
+
+            if status is None and self.statuses:
+                status = self.statuses.pop()
+
+            # Scenario 1: nothing more is highlighted.
+            if ruby is None and status is None:
+                debug_utils.dev_print("Scenario 1: nothing more is highlighted.")
+                break
+
+            # Scenario 2: There are only statuses.
+            if ruby is None:
+                debug_utils.dev_print("Scenario 2: There are only statuses.")
+                # Ignore is here because (surprisingly) mypy can not tell the
+                # only path that leads here requires stat to be non-None.
+                self._highlighted_expression = status.inject(self._highlighted_expression)  # type: ignore[union-attr]
+                status = None
+                continue
+
+            # Scenario 3: There are only rubies.
+            if status is None:
+                debug_utils.dev_print("Scenario 3: There are only rubies.")
+                self._highlighted_expression = ruby.inject(self._highlighted_expression)
+                ruby = None
+                continue
+
+            debug_utils.dev_print(f"current ruby: {repr(ruby)}")
+            debug_utils.dev_print(f"current status: {repr(status)}")
+
+            # Scenario 4: There is no overlap between ruby and status.
+            if ruby.end <= status.start or ruby.start >= status.end:
+                debug_utils.dev_print(
+                    "Scenario 4: There is no overlap between ruby and status."
+                )
+                if ruby.start > status.start:
+                    self._highlighted_expression = ruby.inject(
+                        self._highlighted_expression
+                    )
+                    ruby = None
+                else:
+                    self._highlighted_expression = status.inject(
+                        self._highlighted_expression
+                    )
+                    status = None
+                continue
```

Comment:
Cases 2 and 3 turn out to be specializations of case 4. With this change, 2 and 3 can be deleted and all tests still pass.

```suggestion
            if (
                ruby is None
                or status is None
                or ruby.end <= status.start
                or ruby.start >= status.end
            ):
                debug_utils.dev_print(
                    "Scenario 4: There is no overlap between ruby and status."
                )
                if status is None or (ruby is not None and ruby.start > status.start):
                    self._highlighted_expression = ruby.inject(
                        self._highlighted_expression
                    )
                    ruby = None
                else:
                    self._highlighted_expression = status.inject(
                        self._highlighted_expression
                    )
                    status = None
                continue
```